aptos/contract/aux/sources/stake.move

```diff
@@ -0,0 +1,1399 @@
+        } else {
+            coin::destroy_zero(reward);
+        };
+    }
+
+    /// Modify the incentive pool authority.
+    /// Only the pool's current authority can modify the authority.
+    public entry fun modify_authority<S, R>(
+        sender: &signer,
+        id: u64,
+        new_authority: address
+    ) acquires Pools {
+        assert!(exists<Pools<S, R>>(@aux), E_INCENTIVE_POOL_NOT_FOUND);
+        let pools = borrow_global_mut<Pools<S, R>>(@aux);
+        assert!(table::contains(&mut pools.pools, id), E_INCENTIVE_POOL_NOT_FOUND);
+        let pool = table::borrow_mut(&mut pools.pools, id);
+
+        let sender_addr = signer::address_of(sender);
+        assert!(sender_addr == pool.authority, E_NOT_AUTHORIZED);
+        let now = timestamp::now_microseconds();
+        update_pool(pool, now);
+
+        pool.authority = new_authority;
+
+        event::emit_event<ModifyPoolEvent<S, R>>(
+            &mut pools.modify_pool_events,
+            ModifyPoolEvent {
+                pool_id: id,
+                authority: pool.authority,
+                start_time: pool.start_time,
+                end_time: pool.end_time,
+                reward_remaining: pool.reward_remaining,
+                total_amount_staked: coin::value(&pool.stake),
+                acc_reward_per_share: pool.acc_reward_per_share
+            }
+        );
+    }
+
+    /// Modify incentive pool parameters.
+    /// Only the pool's authority can modify parameters.
+    public entry fun modify_pool<S, R>(
+        sender: &signer,
+        id: u64,
+        reward_amount: u64,
+        reward_increase: bool,
+        time_amount_us: u64,
+        time_increase: bool,
+    ) acquires Pools {
+        assert!(exists<Pools<S, R>>(@aux), E_INCENTIVE_POOL_NOT_FOUND);
+        let pools = borrow_global_mut<Pools<S, R>>(@aux);
+        assert!(table::contains(&mut pools.pools, id), E_INCENTIVE_POOL_NOT_FOUND);
+        let pool = table::borrow_mut(&mut pools.pools, id);
+
+        let sender_addr = signer::address_of(sender);
+        assert!(sender_addr == pool.authority, E_NOT_AUTHORIZED);
+
+        let now = timestamp::now_microseconds();
+        update_pool(pool, now);
+
+        if (reward_amount > 0 && reward_increase) {
+            let reward = coin::withdraw<R>(sender, reward_amount);
+            coin::merge(&mut pool.reward, reward);
+            pool.reward_remaining = pool.reward_remaining + reward_amount;
+        } else if (reward_amount > 0 && !reward_increase) {
+            assert!(pool.reward_remaining >= reward_amount, E_INVALID_REWARD);
+            let reward = coin::extract(&mut pool.reward, reward_amount);
+            if (!coin::is_account_registered<R>(sender_addr)) {
+                coin::register<R>(sender);
+            };
+            coin::deposit(sender_addr, reward);
+            pool.reward_remaining = pool.reward_remaining - reward_amount;
+            if (pool.reward_remaining == 0) {
+                pool.end_time = now;
```

Comment:
What happens if the reward is drained and then the authority adds additional reward? Are timestamps handled correctly in this case?

Reply:
The only ways for the reward to be drained are 1. Reward period ends so all reward is distributed and 2. Authority removes all reward. In both cases the authority may add more reward but the authority needs to set a new duration first. I should probably rearrange this so that it checks for end time modifications before reward modifications so that can be done in a single step.

Reply:
EDIT: actually this can already be done in one step. See https://github.com/aux-exchange/aux-exchange/blob/b81e31e4a7eed3f5efd4fe4e346888d8296acd08/aptos/contract/aux/sources/stake.move#L1403 for an example of this. If the pool's end time has passed (which is required in order for the reward to be drained, either due to its natural reward schedule or an early withdrawal that sets pool's end time to `now`), then the start_time will reset to the time when the new reward is added and the new duration will be added to that start time to get the end time. If an authority attempts to add more reward without updating the duration, the update will now fail due to this check: https://github.com/aux-exchange/aux-exchange/blob/af9d1a7f5537fb32df4b7b84a976f870c2f1044a/aptos/contract/aux/sources/stake.move#L317-L319